File: languages/js/src/entity/authenticity/bjj_signer.ts
import * as proto from "../../bridge/proto/authenticity_entities";
import { LocalKey, ManagedKey } from "../key";
import { Signer } from "./signer";
import { SignerArgs } from "./signer_args";

export class BjjSigner implements Signer {
  alg: proto.SignerAlg;
  args: SignerArgs;

  constructor(
    key: LocalKey | ManagedKey | string,
    options?: { commonName?: string }
  ) {
    this.alg = proto.SignerAlg.BJJ;
    this.args = new SignerArgs(key, options?.commonName);
  }

  public toProto(): proto.Signer {
    return proto.Signer.fromPartial({
      alg: this.alg,
      localKey: this.args.localKey?.toProto(),
      managedKey: this.args.managedKey?.toProto(),
      commonName: this.args.commonName
    });
  }
}

File: languages/go/entity/authenticity/bjj_signer.go
package authenticity

import (
	"github.com/bloock/bloock-sdk-go/v2/entity/key"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
)

type BjjSigner struct {
	Alg  proto.SignerAlg
	Args SignerArgs
}

func NewBjjSigner(args SignerArgs) BjjSigner {
	return BjjSigner{
		Alg:  proto.SignerAlg_BJJ,
		Args: args,
	}
}

func (s BjjSigner) ToProto() *proto.Signer {
	var localKey *proto.LocalKey
	if s.Args.LocalKey != nil {
		localKey = s.Args.LocalKey.ToProto()
	} else if s.Args.PrivateKey != "" {
		localKey = key.LocalKey{
			PrivateKey: s.Args.PrivateKey,
		}.ToProto()
	}

	var managedKey *proto.ManagedKey
	if s.Args.ManagedKey != nil {
		managedKey = s.Args.ManagedKey.ToProto()
	}

	return &proto.Signer{
		Alg:        s.Alg,
		LocalKey:   localKey,
		ManagedKey: managedKey,
		CommonName: s.Args.CommonName,
	}
}